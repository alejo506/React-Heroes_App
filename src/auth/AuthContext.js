//Hacemos la importación de Context
import { createContext } from 'react';

//Este es nuestro context. Lo utilizaremos en HeroesApp.js, se puede utilizar en el index.js pero es mejor dejar nuestro index.js lo más limpio posible, lo hacemos el HeroesApp porque ahí es donde ya comienza a desplegarse toda mi aplicación
export const AuthContext = createContext();


