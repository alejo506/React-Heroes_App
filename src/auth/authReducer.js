import { types } from "../types/types";

// //Este será el estado si nuestro usuario está autenticado, y si no está autenticado no va a tener la propiedad "name" y el logged estará en "false"
// const state = {
//     name: 'Alejandro',
//     logged: true
// }
//Exportamos el reducer
export const authReducer = ( state = {}, action)=>{//El reducer recibe el state elcual es un objeto. La action es la accion que vamos a realizar

    
    
    switch ( action.type ) {//Evaluaremos el action.type
        //Las dos acciones que vamos a realizar es el Login y el Logout
        case types.login:
            
        return{
            //Retornamos todo lo que viene en el action.payload (...action.payload)
            ...action.payload, //Tomamos lo que viene en el payload de la accion y se establece abajo el logged en true, esto es porque si enviamos un name:'Alejandro' este va a ser el payload de la de accion y si recibe dicha propiedad entonces el logged cambia a true por el contrario, si no enviamos el name:'Alejandro' solo se establece la propiedad logged en false(en este caso es logout)
            logged: true //La autenticación del usuario. Si pasa la autenticación va a ser tru
        }
        
        case types.logout:
            
            return { //En el logout no nos interesará nada del payload, solo retornaremos el estamos de la propiedad logged en "false"
                logged: false
            }
            
            default:
                return state; //Retornamos el estado tal y como se encuentra originalmente
            }
            
}