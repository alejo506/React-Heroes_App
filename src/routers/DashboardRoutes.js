import React from 'react'
import { Redirect, Route, Switch } from 'react-router'
import { DcScreen } from '../components/dc/DcScreen'
import { HeroeScreen } from '../components/heroes/HeroeScreen'
import { MarvelScreen } from '../components/marvel/MarvelScreen'
import { SearchScreen } from '../components/search/SearchScreen'
import { Footer } from '../components/ui/Footer'
import { Navbar } from '../components/ui/Navbar'

export const DashboardRoutes = (/*{history}*/) => {
    return (
        <> {/*Este fragment se agrega porque hay más de un elemento html, hay un <Navbar/> y el <div/> */}
         
         <Navbar /*history={history}*/ /> {/*Se pasa el history al Navbar para utilizar la propiedad "history". Solo que en lugar de hacer esto, utilizamos el customHook useHistory en el ui/Navbar.*/}
         
          
            <div className="container mt-2">
                <Switch>
  
                    <Route exact path="/marvel" component={ MarvelScreen } />
                    {/* Este componente va a recibir un argumento por el url ":heroeId" */}
                    <Route exact path="/hero/:heroeId" component={ HeroeScreen } />
                    <Route exact path="/dc" component={ DcScreen } />
                    <Route exact path="/search" component={ SearchScreen } />

                    {/* Si no nos encontramos en alguna de la rutas anteriores(marvel,heroe/:heroeID o dc) entonces nos redireccione a path de marvel */}
                 {/*El Redirect también se puede utilizar para validaciones, por ejemplo, si algo no viene entonces que llame al Redirect  */}
                  <Redirect to="/marvel" />
                 
                </Switch>  
            </div>  

            <Footer/>
        </>
    )
}