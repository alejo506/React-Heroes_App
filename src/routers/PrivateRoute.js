import React from 'react'
import { Redirect, Route } from 'react-router-dom'
import PropTypes from 'prop-types'; //Importamos los Protypes porque necesito obligarme a utilizarlo de manera correcta. Necesitamos al menos al isAuthenticated y el componente 

//Utilizamos las llaves porque vamos a recibir varios elementos en las properties
export const PrivateRoute = ({
    isAuthenticated, //De alguna manera tenemos que saber si está autenticado por lo que creamos una propiedad llamada "isAuthenticated"
    component: Component, //También necesitaremos el componente que la persona quiere renderizar. lo renombramos al componente como "Component" con C mayuscula porque si lo dejamos en minuscula se miraría como un elemento html
    ...rest  //Los demás argumentos como el exact, el path y todo lo demás necesitamos recuperlos y los almacenamos utilizando el operador ...rest(el resto), de esta manera, podremos pasarle dichos argumentos al componente de la manera en la que queramos
}) => {

    //--------------------------------------------------------------------------------------------------------------------------------------
    //193. Recordar la última pagina visitada. 
    //Del rest ocuparemos el location.pathname
    //No es necesario utilizar el customHook para acceder al "location" porque ya lo tenemos en el ..rest

    // console.log(rest.location.pathname); //Esto lo almacenamos en el LocalStorage
    //Almacenamiento del pathname en LocalStorage
    localStorage.setItem('last_Pathname', rest.location.pathname) //Le damos por nombre "last_Pathname". Almacenamos el pathname pero, no almacenamos el query, si quisieramos podemos almacenar el pathname y concatenarlo con el query (el query por ejemplo en el componente Search si busco algo en el input lo que escriba en ese input va a ser el query)
    //Ahora nos vamos al componente Login a implementarlo

    //--------------------------------------------------------------------------------------------------------------------------------------


    return (
        //Regresamos o retornamos un "Route". Vamos a regresar una ruta de "react-router DOM"
        //La ruta a a tener el ..rest(resto de propiedades que estamos recibiendo ahí, el exact, el path y todo lo demás)
        <Route {...rest} //El ...rest tiene almacena el exact, el path y demás cosas. El path, exact y más es enviado a este Route

            //Queremos retornar el componente de manera condicional
            component={(props) => (//El componente se va a llamr con este callback y va a recibir las "props", las props son el history, location y params. Como solo vamos a regresar algo quitamos las llaves {} y ponemos parentesis ()

                //Hacemos la condición con un oprador ternario
                (isAuthenticated)// Si está autenticado regresa el componente al cual el usuario quiere entrar, y le adicionamos con el operador express las "props", recordar que las props son history, location y params
                    ? (<Component {...props} />) // Recordemos que este componente hace referencia al component ubicado en el archivo AppRouter y que ese es component={DashboardRoutes}, eso quiere decir que si está autenticado nos devuelve al DashboardRoutes y que este tiene como ruta principal a /marvel.Como saber cual es el operador express y cual es el rest. La respuesta es que el que está en los argumentos es el ...rest y el que se encuentra en en el Component es el express, en este caso "...props" es el operador express, en conclusion dependiendo del lugar donde se use le cambia el nombre
                    //Component es el componente que queremos renderizar, eso quiere decir que si estamos logueados o autenticados manda todo lo mismo de igual forma (manda las ...props las cuales son el history, location y params). Con esto todo manda lo mismo, es decir todo queda igual.
                    //Si no está autenticado utilizamos el componente Redirect, para redirecionarlo a la pagina de Login
                    : (<Redirect to="/login" />) //Podemos agregar la ruta que la persona quiera, podemos leer el path y muchas cosas y almacenar eso en el LocalStorage o en una cookie, para que cuando el usuario se autentique vaya a la pagina que él quería   
            )}

        />
    )
}

//Los Proptypes nos permite utilizar nuestro componente PriveRoute de manera correcta. 
PrivateRoute.propTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    component: PropTypes.func.isRequired
}


/*Una vez tenemos listo este componente PrivateRoute lo vamos a implementar en AppRouter, en AppRouter sustituimos el Route que contiene nuestro DashboardRoutes por el PrivateRoute para que proteja todas las rutas dentro del DashboardRoutes
Podemos utilizar nuestro PrivateRoute en el DashboarRoutes, pero, no es necesarios cambiar cada Route por PrivateRoute y que con tan solo proteger la Route de DashboardRoutes que se encuentra en AppRoute protejo a todas las que se encuentra en el DashboarRoutes*/