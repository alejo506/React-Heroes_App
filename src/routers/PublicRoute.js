import React from 'react'
import { Redirect, Route } from 'react-router'
import PropTypes from 'prop-types';

// Esta ruta es similar a la de PrivateRoute solo que tiene un cambio

export const PublicRoute = ({ //Se utiliza llaves porque PublicRoute RECIBE por argumento lo definido PublicRoute que se enceuntra en AppRouter
    isAuthenticated,
    component: Component, //componente es renombrado Component en este archivo
    ...rest
}) => {
    return (

        <Route {...rest} // ...rest = pacth, exatc, etc...
         
        component = { (props) => ( //Con esete callback recibimos las props

            //De esta manera si estamos autenticados y alguien pone en el url para ir al Login, no lo va a permitir
            (!isAuthenticated)// Si NO está autenticado mostramos el componente
                ? (<Component { ...props } />) //Este Component recibe lo del component que se ubica en el archivo AppRouter.Muestra el componente que está solicitando únicamente si NO está autenticado. Recordemos que este componente hace referencia al component ubicado en el archivo AppRouter y que ese es component={LoginScreen}, eso quiere decir que si NO está autenticado nos devuelve al Login.El componente recibe todas las props(history, location y params).
                : (<Redirect to="/" />) //Si está autenticado lo enviamos al DashboarRoute y no al Login
             
        
            )}

        
        />
        
    )
}

PublicRoute.propTypes ={
    isAuthenticated: PropTypes.bool.isRequired,
    component: PropTypes.func.isRequired
}