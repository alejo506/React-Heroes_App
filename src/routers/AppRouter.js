import React, { useContext } from "react";
import {
    BrowserRouter as Router,
    HashRouter,
    Switch,
    // Route

} from "react-router-dom";
import { AuthContext } from "../auth/AuthContext";
import { LoginScreen } from "../components/login/LoginScreen";
import { MarvelScreen } from "../components/marvel/MarvelScreen";

import { Navbar } from "../components/ui/Navbar";
import { DashboardRoutes } from "./DashboardRoutes";
import { PrivateRoute } from "./PrivateRoute";
import { PublicRoute } from "./PublicRoute";

export const AppRouter = () => {

    // Nos vamos a components en el navegador y nos dirigimos a HeroesApp y ahí nos daremos cuenta si está autenticado o no, para hacer eso vamos a hacer el useContext
    const {user} = useContext(AuthContext);
    // console.log(user);

    return (
        <HashRouter>
        {/* <Router> */}
            <div>


                <Switch>

                    {/* exact para que la ruta siempre sea exacta */}
                    <PublicRoute
                     exact path="/login" 
                     component={LoginScreen} 
                     isAuthenticated = {user.logged}
                     />

                    {/*No se agrega el exact porque se presentan inconvenientes */}
                    <PrivateRoute 
                    path="/" //El path no lo recibe ni la propiedad isAuthenticated ni el component, lo recibe el operador ..rest(recibe el resto o lo demás, en este caso recibe el path , el exact y otros, en este caso solo recibe el path porque es el unico que se encuetra acá)  
                    component={DashboardRoutes} //Este es el componente que carga si está autenticado, este componente lo recibe el PrivateRoute
                    isAuthenticated= {user.logged} //Enviamos el valor de la propiedad logged para saber si está o no autenticado, este valor lo recibe la propiedad isAutheticated en el componente PrivateRoute.js
                    /> {/*Implementamos nuestro PrivateRoute. Por ejemplo, el path que se encuentra aquí no está como isAutheticated(autenticado) ni component(Componente), entonces el path cae en el ...rest(todo esto en el archivo PrivateRoute.js)*/}
               
                </Switch>
            </div>
        {/* </Router> */}
                    </HashRouter>
    )
}
