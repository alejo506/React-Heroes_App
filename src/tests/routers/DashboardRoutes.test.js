
import React from 'react';
import { mount } from 'enzyme';

import { MemoryRouter } from 'react-router-dom';
import { DashboardRoutes } from '../../routers/DashboardRoutes';
import { AuthContext } from '../../auth/AuthContext';


describe('Pruebas en el <DashboardRoutes', () => {

    //Hacemos el uso del Context
    //Si estamos logueados necesitaremos el objeto user y hacer la simulacion del dispatch
    const contextValue = {
        dispatch: jest.fn(),
        user: {
            logged: true,
            name: 'Natasha'
        }
    }

    test('debe de mostrarse correctamente ', () => {

        const wrapper = mount(
            // Tenemos que utilizar el Memory Router porque estamos trabajando con Route. Errores: You should not use <Switch> outside a <Router>. Otro error: You should not use <Link> outside a <Router>
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter>
                    <DashboardRoutes />
                </MemoryRouter>
            </AuthContext.Provider>

        );

        // console.log(wrapper.html());

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.textInfo').text().trim()).toBe('Natasha');
    })

})
