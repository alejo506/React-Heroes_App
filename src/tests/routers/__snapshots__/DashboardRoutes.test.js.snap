// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <DashboardRoutes debe de mostrarse correctamente  1`] = `
Array [
  " ",
  <nav
    className="navbar navbar-expand-lg navbar-dark bg-dark"
  >
    <div
      className="container-fluid"
    >
      <div
        className="nav-item "
      >
        <a
          className="navbar-brand nav-link"
          href="/"
          onClick={[Function]}
          style={
            Object {
              "marginLeft": 15,
            }
          }
        >
          Asociaciones
        </a>
      </div>
      <button
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Toggle navigation"
        className="navbar-toggler"
        data-bs-target="#navbarSupportedContent"
        data-bs-toggle="collapse"
        type="button"
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="collapse navbar-collapse"
        id="navbarSupportedContent"
      >
        <ul
          className="navbar-nav me-auto mb-1 mb-lg-0"
        >
          <div
            className="nav-item "
          >
            <a
              aria-current="page"
              className="nav-link active"
              href="/marvel"
              onClick={[Function]}
              style={Object {}}
            >
              Marvel
            </a>
          </div>
          <div
            className="nav-item "
          >
            <a
              aria-current={null}
              className="nav-link"
              href="/dc"
              onClick={[Function]}
            >
              DC
            </a>
          </div>
          <div
            className="nav-item"
          >
            <a
              aria-current={null}
              className="nav-link"
              href="/search"
              onClick={[Function]}
            >
              Search
            </a>
          </div>
        </ul>
        <div
          className="nav-item dropdown "
        >
          <a
            aria-expanded="false"
            className="navbar-brand nav-link dropdown-toggle"
            data-bs-toggle="dropdown"
            id="navbarDarkDropdownMenuLink"
            role="button"
          >
            <lord-icon
              colors="primary:#e4e4e4,secondary:#e4e4e4"
              src="https://cdn.lordicon.com/dxjqoygy.json"
              stroke="60"
              style={
                Object {
                  "height": 38,
                  "width": 38,
                }
              }
              trigger="hover"
            />
             
          </a>
          <ul
            aria-labelledby="navbarDarkDropdownMenuLink"
            className="dropdown-menu dropdown-menu-dark"
          >
            <span
              className="ms-3 textInfo"
              style={
                Object {
                  "color": "#5cb85c",
                }
              }
            >
               
              Natasha
               
            </span>
            <button
              className="dropdown-item"
              onClick={[Function]}
            >
              Logout
            </button>
          </ul>
        </div>
      </div>
    </div>
  </nav>,
  " ",
  <div
    className="container mt-2"
  >
    <div>
      <div
        className="contenedorPadre "
      >
        <div
          className="marvel-logo animate__animated animate__pulse"
        >
           
          <img
            className="card-img"
          />
           
        </div>
      </div>
      <hr />
      <div
        className="container"
      >
        <div
          className=" row row-cols-1 row-cols-md-2 g-4 g-lg-3 row-cols-xl-4 g-xl-auto row-cols-xs-2 g-xs-2 animate__animated animate__fadeIn"
        >
          <a
            href="/hero/marvel-spider"
            onClick={[Function]}
          >
             
            <div
              className="card my-card "
            >
              <img
                alt="Spider Man"
                className="card-img"
              />
              <div
                className="profile-name"
              >
                Spider Man
              </div>
              <div
                className="profile-position"
              >
                Peter Parker
              </div>
              <div
                className="profile-overview"
              >
                <div
                  className="profile-overview"
                >
                  <h3>
                    Marvel Comics
                  </h3>
                  <p>
                    Primera aparición: 
                    <br />
                    Amazing Fantasy #15
                  </p>
                </div>
              </div>
            </div>
          </a>
          <a
            href="/hero/marvel-captain"
            onClick={[Function]}
          >
             
            <div
              className="card my-card "
            >
              <img
                alt="Captain America"
                className="card-img"
              />
              <div
                className="profile-name"
              >
                Captain America
              </div>
              <div
                className="profile-position"
              >
                Steve Rogers
              </div>
              <div
                className="profile-overview"
              >
                <div
                  className="profile-overview"
                >
                  <h3>
                    Marvel Comics
                  </h3>
                  <p>
                    Primera aparición: 
                    <br />
                    Captain America Comics #1
                  </p>
                </div>
              </div>
            </div>
          </a>
          <a
            href="/hero/marvel-iron"
            onClick={[Function]}
          >
             
            <div
              className="card my-card "
            >
              <img
                alt="Iron Man"
                className="card-img"
              />
              <div
                className="profile-name"
              >
                Iron Man
              </div>
              <div
                className="profile-position"
              >
                Tony Stark
              </div>
              <div
                className="profile-overview"
              >
                <div
                  className="profile-overview"
                >
                  <h3>
                    Marvel Comics
                  </h3>
                  <p>
                    Primera aparición: 
                    <br />
                    Tales of Suspense #39
                  </p>
                </div>
              </div>
            </div>
          </a>
          <a
            href="/hero/marvel-thor"
            onClick={[Function]}
          >
             
            <div
              className="card my-card "
            >
              <img
                alt="Thor"
                className="card-img"
              />
              <div
                className="profile-name"
              >
                Thor
              </div>
              <div
                className="profile-position"
              >
                Thor Odinson
              </div>
              <div
                className="profile-overview"
              >
                <div
                  className="profile-overview"
                >
                  <h3>
                    Marvel Comics
                  </h3>
                  <p>
                    Primera aparición: 
                    <br />
                    Journey into Myster #83
                  </p>
                </div>
              </div>
            </div>
          </a>
          <a
            href="/hero/marvel-hulk"
            onClick={[Function]}
          >
             
            <div
              className="card my-card "
            >
              <img
                alt="Hulk"
                className="card-img"
              />
              <div
                className="profile-name"
              >
                Hulk
              </div>
              <div
                className="profile-position"
              >
                Bruce Banner
              </div>
              <div
                className="profile-overview"
              >
                <div
                  className="profile-overview"
                >
                  <h3>
                    Marvel Comics
                  </h3>
                  <p>
                    Primera aparición: 
                    <br />
                    The Incredible Hulk #1
                  </p>
                </div>
              </div>
            </div>
          </a>
          <a
            href="/hero/marvel-wolverine"
            onClick={[Function]}
          >
             
            <div
              className="card my-card "
            >
              <img
                alt="Wolverine"
                className="card-img"
              />
              <div
                className="profile-name"
              >
                Wolverine
              </div>
              <div
                className="profile-position"
              >
                James Howlett
              </div>
              <div
                className="profile-overview"
              >
                <div
                  className="profile-overview"
                >
                  <h3>
                    Marvel Comics
                  </h3>
                  <p>
                    Primera aparición: 
                    <br />
                    The Incredible Hulk #180
                  </p>
                </div>
              </div>
            </div>
          </a>
          <a
            href="/hero/marvel-daredevil"
            onClick={[Function]}
          >
             
            <div
              className="card my-card "
            >
              <img
                alt="Daredevil"
                className="card-img"
              />
              <div
                className="profile-name"
              >
                Daredevil
              </div>
              <div
                className="profile-position"
              >
                Matthew Michael Murdock
              </div>
              <div
                className="profile-overview"
              >
                <div
                  className="profile-overview"
                >
                  <h3>
                    Marvel Comics
                  </h3>
                  <p>
                    Primera aparición: 
                    <br />
                    Daredevil #1
                  </p>
                </div>
              </div>
            </div>
          </a>
          <a
            href="/hero/marvel-hawkeye"
            onClick={[Function]}
          >
             
            <div
              className="card my-card "
            >
              <img
                alt="Hawkeye"
                className="card-img"
              />
              <div
                className="profile-name"
              >
                Hawkeye
              </div>
              <div
                className="profile-position"
              >
                Clinton Francis Barton
              </div>
              <div
                className="profile-overview"
              >
                <div
                  className="profile-overview"
                >
                  <h3>
                    Marvel Comics
                  </h3>
                  <p>
                    Primera aparición: 
                    <br />
                    Tales of Suspense #57
                  </p>
                </div>
              </div>
            </div>
          </a>
          <a
            href="/hero/marvel-cyclops"
            onClick={[Function]}
          >
             
            <div
              className="card my-card "
            >
              <img
                alt="Cyclops"
                className="card-img"
              />
              <div
                className="profile-name"
              >
                Cyclops
              </div>
              <div
                className="profile-position"
              >
                Scott Summers
              </div>
              <div
                className="profile-overview"
              >
                <div
                  className="profile-overview"
                >
                  <h3>
                    Marvel Comics
                  </h3>
                  <p>
                    Primera aparición: 
                    <br />
                    X-Men #1
                  </p>
                </div>
              </div>
            </div>
          </a>
          <a
            href="/hero/marvel-silver"
            onClick={[Function]}
          >
             
            <div
              className="card my-card "
            >
              <img
                alt="Silver Surfer"
                className="card-img"
              />
              <div
                className="profile-name"
              >
                Silver Surfer
              </div>
              <div
                className="profile-position"
              >
                Norrin Radd
              </div>
              <div
                className="profile-overview"
              >
                <div
                  className="profile-overview"
                >
                  <h3>
                    Marvel Comics
                  </h3>
                  <p>
                    Primera aparición: 
                    <br />
                    The Fantastic Four #48
                  </p>
                </div>
              </div>
            </div>
          </a>
        </div>
      </div>
       
    </div>
  </div>,
  <footer
    className="bg-dark text-center text-white"
    style={
      Object {
        "bottom": 0,
        "marginTop": 233,
      }
    }
  >
    <div
      className="text-center p-3"
      style={
        Object {
          "backgroundColor": "#000000",
          "opacity": 0.2,
        }
      }
    >
      © 2021 Copyright:
      <a
        className="text-white ms-1"
        href="https://www.youtube.com/watch?v=LXb3EKWsInQ"
      >
         Alejandro CM
      </a>
    </div>
  </footer>,
]
`;
